docs/design/2020-09-28-cluster-read-only-mode.md

```diff
@@ -0,0 +1,243 @@
+    *   Even if the user’s query actually did not update any value, the query might still be denied (tested with MySQL), such as \
+insert into t (select * from an_empty_table)
+    *   This implies that we could just block the SQL by its type, namely the following kinds of query will be banned if read only is turned on:
+        *   INSERT
+        *   UPDATE
+        *   SELECT * FOR UPDATE
+        *   LOCK TABLE FOR WRITE
+        *   All DDL
+        *   EXPLAIN ANALYZE where the analyzed SQL might update the data.
+    *   All other SQL will be allowed and will be executed just as normal.
+*   Users with `SuperPriv` privilege level are not affected by the read only mode; they can still write data to the cluster.
+*   User identified as a "replication secondary" is also freed from the restraint of this option
+    *   A new privilege level called `ReplicationWrite` will be added, in order to allow data migration tools like TiCDC to write data even if the cluster is read only.
+    *   Because MySQL’s replication applier is inside the MySQL server, therefore it does not need such a privilege level. For TiCDC and tools alike, the SQL interface is used to apply data, therefore a new privilege level would be a good idea to allow such change.
+*   Once we support [temporary tables](https://github.com/pingcap/tidb/issues/1248) in the future, updates to temporary tables should also be allowed, but we could ignore this behavior for now, this is described in [MySQL’s documentation](https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_read_only) as well.
+
+
+### Enabling Attempt
+
+As with MySQL, the read_only mode can be enabled at start using a command option or interactively without a restart. In our case, we might just make it a global variable, and it should be set directly from the SQL interface. We only allow `SET GLOBAL read_only=true` to enable the read only mode, there is no other way to enable it, and it can’t be set from a session level.
+
+
+#### Case 1, clients have transactions committing
+
+
+|server 1|server 2|
+|--- |--- |
+|client 1|client 2|
+||begin tx|
+||working|
+||working|
+||committing|
+|turn on read_only|committing|
+|block|committing|
+|block|committed|
+|read_only turned on||
+
+
+
+If there are transactions that are committing (i.e. started to commit before enabling attempt), the enabling attempt will be blocked until the commit is finished. This behavior is inline with MySQL’s, according to [MySQL’s documentation](https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_read_only). If a transaction has not reached commit point, once the enabling attempt is initiated, the transaction will fail at commit point.
+
+
+#### Case 2, client who attempts to turn on `read_only` has write lock held
+
+
+|server 1| server 1|
+|--- |--- |
+|client 1, connection 1|client 1, connection 2|
+|lock table write||
+||turn on read_only|
+||read_only turned on|
```

Comment:
Because the user has SuperPriv, and she can write to table even if read only is turned on, therefore the write lock will not prevent the turning on of read only.